rslib/src/import_export/text/import.rs

```diff
@@ -0,0 +1,239 @@
+            let mut notetype = foreign.into_native();
+            notetype.usn = self.usn;
+            self.col
+                .add_notetype_inner(&mut notetype, self.usn, false)?;
+        }
+        Ok(())
+    }
+    fn notetype_for_note(&mut self, note: &ForeignNote) -> Result<Option<Arc<Notetype>>> {
+        Ok(if let Some(nt) = self.notetypes.get(&note.notetype) {
+            nt.clone()
+        } else {
+            let nt = self.col.notetype_for_string(&note.notetype)?;
+            self.notetypes.insert(note.notetype.clone(), nt.clone());
+            nt
+        })
+    }
+
+    fn deck_id_for_note(&mut self, note: &ForeignNote) -> Result<Option<DeckId>> {
+        Ok(if let Some(did) = self.deck_ids.get(&note.deck) {
+            *did
+        } else {
+            let did = self.col.deck_id_for_string(&note.deck)?;
+            self.deck_ids.insert(note.deck.clone(), did);
+            did
+        })
+    }
+
+    fn import_foreign_notes(&mut self, notes: Vec<ForeignNote>) -> Result<NoteLog> {
+        let mut log = NoteLog::default();
+        for foreign in notes {
+            if let Some(notetype) = self.notetype_for_note(&foreign)? {
+                if let Some(deck_id) = self.deck_id_for_note(&foreign)? {
+                    let log_note = self.import_foreign_note(foreign, notetype, deck_id)?;
+                    log.new.push(log_note);
+                }
+            }
+        }
+        Ok(log)
+    }
+
+    fn import_foreign_note(
+        &mut self,
+        foreign: ForeignNote,
+        notetype: Arc<Notetype>,
+        deck_id: DeckId,
+    ) -> Result<LogNote> {
+        let today = self.col.timing_today()?.days_elapsed;
+        let (mut note, mut cards) = foreign.into_native(&notetype, deck_id, today);
+        self.import_note(&mut note, &notetype)?;
+        self.import_cards(&mut cards, note.id)?;
+        self.generate_missing_cards(notetype, deck_id, &note)?;
+        Ok(note.into_log_note())
+    }
+
+    fn import_note(&mut self, note: &mut Note, notetype: &Notetype) -> Result<()> {
+        self.col.canonify_note_tags(note, self.usn)?;
+        note.prepare_for_update(notetype, self.normalize_notes)?;
+        note.usn = self.usn;
+        self.col.add_note_only_undoable(note)
+    }
+
+    fn import_cards(&mut self, cards: &mut [Card], note_id: NoteId) -> Result<()> {
+        for card in cards {
+            card.note_id = note_id;
+            self.col.add_card(card)?;
+        }
+        Ok(())
+    }
+
+    fn generate_missing_cards(
+        &mut self,
+        notetype: Arc<Notetype>,
+        deck_id: DeckId,
+        note: &Note,
+    ) -> Result<()> {
+        let card_gen_context = self
+            .card_gen_ctxs
+            .entry((notetype.id, deck_id))
+            .or_insert_with(|| CardGenContext::new(notetype, Some(deck_id), self.usn));
+        self.col
+            .generate_cards_for_existing_note(card_gen_context, note)
+    }
+}
+
+pub(super) trait NotetypeForString {
```

Comment:
I've tried something new here, what do you think? Implementing the required Collection API in traits allows us to setup a mock for testing, which is cheaper to build, and whose state is easier to control. It it also helps a bit with namespaces.
The main motivation was to avoid setting up a whole collection for every unit test. That always felt overkill, but I didn't do any serious benchmarking. Do you think it's worth the addtional boilerplate?

Reply:
Is setting up a collection for each test that requires one such a big deal? Many of our unit tests don't require a collection, and the total runtime of all our Rust tests is only 0.6s currently:

```
[dae] code/desktop/anki [main]  % bazel run rslib:anki_tests
...
test result: ok. 155 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.60s
```

I don't mind such an approach if we need it, but I just wonder whether we really need it :-)

Reply:
You're right, it doesn't seem to be necessary. 🙂 Tests are dwarfed by compilation time. Maybe it would be helpful in cases where a more exotic collection state is required for testing.